packages/libs/deer/src/lib.rs

```diff
@@ -6,4 +6,212 @@
+    ///
+    /// This type should signal missing or empty.
+    ///
+    /// Some formats do not have this concept and should in that case always error out, formats like
+    /// `JSON`, which have multiple representations for `null` should return from this function!
+    ///
+    /// # Errors
+    ///
+    /// Current value is not of type null
+    fn null(self) -> Result<(), Self::Error>;
+
+    /// Deserialize a [`bool`] value.
+    ///
+    /// Some formats might not have this concept and should always error out if that is the case.
+    ///
+    /// > **Hint**: Do not try to coerce values like `1` or `0` to booleans, this is highly
+    /// > discouraged. The [`Deserialize`] implementation or a future version of `deer` should do
+    /// > that instead.
+    ///
+    /// # Errors
+    ///
+    /// Current value is not of type bool
+    fn bool(self) -> Result<bool, Self::Error>;
+
+    /// Deserialize a [`Number`] value.
+    ///
+    /// This is a super-type of primitive types used by Rust, this is due to the fact that some
+    /// formats (like `JSON`) do not have the concept of multiple types that represent
+    /// integers/floats.
+    ///
+    /// This method also enables a default trait implementation for all primitives types.
+    ///
+    /// > **Hint**: Do not try to coerce values like `"1"`, `"0"`, `true`, `false` to numbers, this
+    /// > is highly discouraged. The [`Deserialize`] implementation of a future version of `deer`
+    /// > would implement that functionality.
+    ///
+    /// # Errors
+    ///
+    /// Current value is not of type number
+    fn number(self) -> Result<Number, Self::Error>;
+
+    /// Deserialize a [`String`] value.
+    ///
+    /// `deer` (in contrast to [`serde`]) does not support `&str`, this is because `deer` only
+    /// handles owned data, this comes with some overhead, but enables easier implementation.
+    ///
+    /// > **Hint**: Do not try to cerce values to string, this is highly discouraged and only
+    /// > [`Deserialize`] or a future version of `deer` should implement that functionality, if
+    /// > desired.
+    ///
+    /// # Errors
+    ///
+    /// Current value is not of type string
+    ///
+    /// [`serde`]: https://serde.rs/
+    fn string(self) -> Result<String, Self::Error>;
+
+    /// Deserialize a `Array`
+    ///
+    /// This should not directly deserialize into a `Vec<T>`, but return a type that implements
+    /// `ArrayAccess`, this type then iterates over all values.
+    ///
+    /// # Errors
+    ///
+    /// Current value is not of type array
+    fn array(self) -> Result<Self::Array, Self::Error>;
+
+    /// Deserialize a `Object`
+    ///
+    /// This should not directly deserialize into a object, but instead return a type that
+    /// implements `ObjectAccess`, this type will then go through all entries.
+    ///
+    /// # Errors
+    ///
+    /// Current value is not of type object
+    fn object(self) -> Result<Self::Object, Self::Error>;
```

Comment:
I don't see, why `object` and `array` should return accessor methods. It feels inconsistent with the rest of the methods. Why deserializing this object
```json
{
    "int": "not-a-number",
    "string": 10,
}
```
cannot return `Err` directly with two errors?

Reply:
`Self::Object` is supposed to be a trait that is going to be implemented:

```rust
pub trait ObjectAccess {
	type Error: Error;
	type Deserializer: Deserializer;

	fn entry(&mut self) -> Result<Self::Deserializer, Self::Error>
	fn finish(self) -> Result<(), Self::Error>
}
impl IntoIterator for ObjectAccess {}
```

Reply:
I removed it for now as to not bloat the PR, but can add it back

Reply:
added frame for `ObjectAccess` and `ArrayAccess` in [c8a8e45](https://github.com/hashintel/hash/pull/1170/commits/c8a8e452b4d72c95b18b136f85172f35768d77b8)

Reply:
The question was mainly about if this is needed at all.
Discussed offline: Will keep both, `ObjectAccess` and `ArrayAccess` as this is similar to `serde`'s approach. We can revisit this at a later point.